File: julia/src/AudioFile.jl
# File: Sine.jl
# Project: src
# Created Date: 14/06/2022
# Author: Shun Suzuki
# -----
# Last Modified: 14/06/2022
# Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
# -----
# Copyright (c) 2022 Shun Suzuki. All rights reserved.
# 


mutable struct RawPCM
    _m::Modulation
    _header_ptr
    get_sampling_frequency_division
    set_sampling_frequency_division
    get_sampling_frequency
    function RawPCM(filename::String, sampling_freq::Float64, sampling_freq_div)
        chandle = Ref(Ptr{Cvoid}(0))
        autd3capi_modulation_audio_file.autd_modulation_raw_pcm(chandle, filename, sampling_freq, UInt32(sampling_freq_div))
        m = Modulation(chandle[])
        s = new(m, chandle[])
        s.get_sampling_frequency_division = m.get_sampling_frequency_division
        s.set_sampling_frequency_division = m.set_sampling_frequency_division
        s.get_sampling_frequency = m.get_sampling_frequency
        s
    end
end

mutable struct Wav
    _m::Modulation
    _header_ptr
    get_sampling_frequency_division
    set_sampling_frequency_division
    get_sampling_frequency
    function Wav(filename::String, sampling_freq_div)
        chandle = Ref(Ptr{Cvoid}(0))
        autd3capi_modulation_audio_file.autd_modulation_wav(chandle, filename, UInt32(sampling_freq_div))
        m = Modulation(chandle[])
        s = new(m, chandle[])
        s.get_sampling_frequency_division = m.get_sampling_frequency_division
        s.set_sampling_frequency_division = m.set_sampling_frequency_division
        s.get_sampling_frequency = m.get_sampling_frequency
        s
    end
end

File: nim/src/autd3/audio_file.nim
# File: emulator.nim
# Project: autd3
# Created Date: 13/06/2022
# Author: Shun Suzuki
# -----
# Last Modified: 13/06/2022
# Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
# -----
# Copyright (c) 2022 Shun Suzuki. All rights reserved.
#


import native_methods/autd3capi_modulation_audio_file
import modulation

type Wav* = object of Modulation

proc initWav*(path: string, freqDiv: uint32): Wav =
  AUTDModulationWav(result.p.addr, path, freqDiv)

type RawPCM* = object of Modulation

proc initRawPCM*(path: string, samplingFreq: float64, freqDiv: uint32): RawPCM =
  AUTDModulationRawPCM(result.p.addr, path, samplingFreq, freqDiv)
